python/hsfs/feature_group.py

```diff
@@ -2837,3 +2832,158 @@ def feature_store_name(self):
     def feature_store_id(self):
         """Id of the feature store in which the feature group is located."""
         return self._feature_store_id
+
+
+class SpineGroup(FeatureGroupBase):
+    SPINE_GROUP = "ON_DEMAND_FEATURE_GROUP"
+    ENTITY_TYPE = "featuregroups"
+
+    def __init__(
+        self,
+        storage_connector=None,
+        query=None,
+        data_format=None,
+        path=None,
+        options={},
+        name=None,
+        version=None,
+        description=None,
+        primary_key=None,
+        featurestore_id=None,
+        featurestore_name=None,
+        created=None,
+        creator=None,
+        id=None,
+        features=None,
+        location=None,
+        statistics_config=None,
+        event_time=None,
+        expectation_suite=None,
+        online_enabled=False,
+        href=None,
+        online_topic_name=None,
+        spine=True,
+        dataframe="spine",
+    ):
+        super().__init__(
+            featurestore_id,
+            location,
+            event_time=event_time,
+            online_enabled=online_enabled,
+            id=id,
+            expectation_suite=expectation_suite,
+            online_topic_name=online_topic_name,
+        )
+
+        self._feature_store_name = featurestore_name
+        self._description = description
+        self._created = created
+        self._creator = user.User.from_response_json(creator)
+        self._version = version
+        self._name = name
+
+        # use setter to convert to default dataframe type for engine
+        self.dataframe = dataframe
+
+        self._features = [
+            feature.Feature.from_response_json(feat) if isinstance(feat, dict) else feat
+            for feat in (features or [])
+        ]
+
+        self._feature_group_engine = spine_group_engine.SpineGroupEngine(
+            featurestore_id
+        )
+
+        if self._id:
+            # Got from Hopsworks, deserialize features and storage connector
+            self._features = (
+                [
+                    feature.Feature.from_response_json(feat)
+                    if isinstance(feat, dict)
+                    else feat
+                    for feat in features
+                ]
+                if features
+                else None
+            )
+            self.primary_key = (
+                [feat.name for feat in self._features if feat.primary is True]
+                if self._features
+                else []
+            )
+            self.statistics_config = statistics_config
+        else:
+            self.primary_key = primary_key
+            self.statistics_config = statistics_config
+            self._features = features
+
+        self._href = href
+
+    def _save(self):
+        """Persist the metadata for this external feature group.
```

Comment:
external feature group -> spine group

Reply:
it's a private method, not to be used by users, but fixed anyway